File: njli/platform/cmake.in/ldoc.in/PhysicsConstraintConeTwist.lua

----
-- @file PhysicsConstraintConeTwist


---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:calculateSerializeBufferSize()
end

---- Brief description.
-- @author James Folk, 16-02-11 15:02:15
-- <#Description#>
-- @param dataBuffer <#dataBuffer description#>
-- @param btSerializer <#btSerializer description#>
-- @return 
function PhysicsConstraintConeTwist:serialize(dataBuffer, btSerializer)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getClassName()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getType()
end

---- Brief description.
-- <#Description#>
-- @param nodeA <#nodeA description#>
-- @param rbAFrame <#rbAFrame description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setNodes(nodeA, rbAFrame)
end

---- Brief description.
-- <#Description#>
-- @param nodeA <#nodeA description#>
-- @param nodeB <#nodeB description#>
-- @param rbAFrame <#rbAFrame description#>
-- @param rbBFrame <#rbBFrame description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setNodes(nodeA, nodeB, rbAFrame, rbBFrame)
end

---- Brief description.
-- <#Description#>
-- @param angularOnly <#angularOnly description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setAngularOnly(angularOnly)
end

---- Brief description.
-- <#Description#>
-- @param limitIndex <#limitIndex description#>
-- @param limitValue <#limitValue description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setLimit(limitIndex, limitValue)
end

---- Brief description.
-- <#Description#>
-- @param _swingSpan1 <#_swingSpan1 description#>
-- @param _swingSpan2 <#_swingSpan2 description#>
-- @param _twistSpan <#_twistSpan description#>
-- @param _softness <#.f description#>
-- @param _biasFactor <#.3f description#>
-- @param _relaxationFactor <#.0f description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setLimit(_swingSpan1, _swingSpan2, _twistSpan, _softness, _biasFactor, _relaxationFactor)
end

---- Brief description.
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getAFrame()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getBFrame()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getSolveTwistLimit()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getSolveSwingLimit()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getTwistLimitSign()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getSwingSpan1()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getSwingSpan2()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getTwistSpan()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getTwistAngle()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:isPastSwingLimit()
end

---- Brief description.
-- <#Description#>
-- @param damping <#damping description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setDamping(damping)
end

---- Brief description.
-- <#Description#>
-- @param b <#b description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:enableMotor(b)
end

---- Brief description.
-- <#Description#>
-- @param maxMotorImpulse <#maxMotorImpulse description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setMaxMotorImpulse(maxMotorImpulse)
end

---- Brief description.
-- <#Description#>
-- @param maxMotorImpulse <#maxMotorImpulse description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setMaxMotorImpulseNormalized(maxMotorImpulse)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getFixThresh()
end

---- Brief description.
-- <#Description#>
-- @param fixThresh <#fixThresh description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setFixThresh(fixThresh)
end

---- Brief description.
-- <#Description#>
-- @param q <#q description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setMotorTarget(q)
end

---- Brief description.
-- <#Description#>
-- @param q <#q description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setMotorTargetInConstraintSpace(q)
end

---- Brief description.
-- <#Description#>
-- @param fAngleInRadians <#fAngleInRadians description#>
-- @param fLength <#fLength description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getPointForAngle(fAngleInRadians, fLength)
end

---- Brief description.
-- <#Description#>
-- @param frameA <#frameA description#>
-- @param frameB <#frameB description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:setFrames(frameA, frameB)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getFrameOffsetA()
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function PhysicsConstraintConeTwist:getFrameOffsetB()
end

---- Brief description.
-- <#Description#>
-- @param size <#size description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.createArray(size)
end

---- Brief description.
-- <#Description#>
-- @param array <#array description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.destroyArray(array)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.create()
end

---- Brief description.
-- <#Description#>
-- @param builder <#builder description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.create(builder)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.clone(object)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.copy(object)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.destroy(object)
end

---- Brief description.
-- <#Description#>
-- @param object <#object description#>
-- @param L <#L description#>
-- @param stack_index <#stack_index description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.load(object, L, stack_index)
end

---- Brief description.
-- <#Description#>
-- @return <#return value description#>
function NJLI.PhysicsConstraintConeTwist.type()
end

File: njli/physics/constraint/PhysicsConstraintConeTwist.cpp
//
//  PhysicsConstraintConeTwist.cpp
//  JLIGameEngineTest
//
//  Created by James Folk on 6/19/15.
//  Copyright (c) 2015 James Folk. All rights reserved.
//

#pragma mark Implementation

#include "PhysicsConstraintConeTwist.h"
#include "JLIFactoryTypes.h"
#include "PhysicsConstraintConeTwistBuilder.h"
#include "World.h"
#define TAG "PhysicsConstraintConeTwist.cpp"

#include "PhysicsBody.h"
#include "PhysicsWorld.h"
#include "btConeTwistConstraint.h"

#define FORMATSTRING "{\"PhysicsConstraint\":[]}"
#include "JsonJLI.h"
#include "btPrint.h"

namespace njli
{
  PhysicsConstraintConeTwist::PhysicsConstraintConeTwist()
      : PhysicsConstraint(), m_btConeTwistConstraint(NULL)
  {
  }

  PhysicsConstraintConeTwist::PhysicsConstraintConeTwist(
      const AbstractBuilder &builder)
      : PhysicsConstraint(builder), m_btConeTwistConstraint(NULL)
  {
  }

  PhysicsConstraintConeTwist::PhysicsConstraintConeTwist(
      const PhysicsConstraintConeTwist &copy)
      : PhysicsConstraint(copy), m_btConeTwistConstraint(NULL)
  {
  }

  PhysicsConstraintConeTwist::~PhysicsConstraintConeTwist()
  {
    removeConstraint();
  }

  PhysicsConstraintConeTwist &PhysicsConstraintConeTwist::
  operator=(const PhysicsConstraintConeTwist &rhs)
  {
    if (this != &rhs)
      {
      }
    return *this;
  }

  s32 PhysicsConstraintConeTwist::calculateSerializeBufferSize() const
  {
    // TODO: calculateSerializeBufferSize
    return 0;
  }

  void PhysicsConstraintConeTwist::serialize(void *dataBuffer,
                                             btSerializer *serializer) const
  {
    // TODO: serialize
  }

  const char *PhysicsConstraintConeTwist::getClassName() const
  {
    return "PhysicsConstraintConeTwist";
  }

  s32 PhysicsConstraintConeTwist::getType() const
  {
    return PhysicsConstraintConeTwist::type();
  }

  PhysicsConstraintConeTwist::operator std::string() const
  {
    return njli::JsonJLI::parse(string_format("%s", FORMATSTRING));
  }

  PhysicsConstraintConeTwist **
  PhysicsConstraintConeTwist::createArray(const u32 size)
  {
    return (PhysicsConstraintConeTwist **)World::getInstance()
        ->getWorldFactory()
        ->createArray(PhysicsConstraintConeTwist::type(), size);
  }

  void
  PhysicsConstraintConeTwist::destroyArray(PhysicsConstraintConeTwist **array,
                                           const u32 size)
  {
    World::getInstance()->getWorldFactory()->destroyArray(
        (AbstractFactoryObject **)array, size);
  }

  PhysicsConstraintConeTwist *PhysicsConstraintConeTwist::create()
  {
    return dynamic_cast<PhysicsConstraintConeTwist *>(
        World::getInstance()->getWorldFactory()->create(
            PhysicsConstraintConeTwist::type()));
  }

  PhysicsConstraintConeTwist *PhysicsConstraintConeTwist::create(
      const PhysicsConstraintConeTwistBuilder &builder)
  {
    AbstractBuilder *b = (AbstractBuilder *)&builder;

    return dynamic_cast<PhysicsConstraintConeTwist *>(
        World::getInstance()->getWorldFactory()->create(*b));
  }

  PhysicsConstraintConeTwist *
  PhysicsConstraintConeTwist::clone(const PhysicsConstraintConeTwist &object)
  {
    return dynamic_cast<PhysicsConstraintConeTwist *>(
        World::getInstance()->getWorldFactory()->clone(object, false));
  }

  PhysicsConstraintConeTwist *
  PhysicsConstraintConeTwist::copy(const PhysicsConstraintConeTwist &object)
  {
    return dynamic_cast<PhysicsConstraintConeTwist *>(
        World::getInstance()->getWorldFactory()->clone(object, true));
  }

  void PhysicsConstraintConeTwist::destroy(PhysicsConstraintConeTwist *object)
  {
    if (object)
      {
        PhysicsConstraint::destroy(object);
      }
  }

  void PhysicsConstraintConeTwist::load(PhysicsConstraintConeTwist &object,
                                        lua_State *L, int index)
  {
    // Push another reference to the table on top of the stack (so we know
    // where it is, and this function can work for negative, positive and
    // pseudo indices
    lua_pushvalue(L, index);
    // stack now contains: -1 => table
    lua_pushnil(L);
    // stack now contains: -1 => nil; -2 => table
    while (lua_next(L, -2))
      {
        // stack now contains: -1 => value; -2 => key; -3 => table
        // copy the key so that lua_tostring does not modify the original
        lua_pushvalue(L, -2);
        // stack now contains: -1 => key; -2 => value; -3 => key; -4 => table
        const char *key = lua_tostring(L, -1);
        //            const char *value = lua_tostring(L, -2);
        if (lua_istable(L, -2))
          {
            PhysicsConstraintConeTwist::load(object, L, -2);
          }
        else
          {
            if (lua_isnumber(L, index))
              {
                double number = lua_tonumber(L, index);
                printf("%s => %f\n", key, number);
              }
            else if (lua_isstring(L, index))
              {
                const char *v = lua_tostring(L, index);
                printf("%s => %s\n", key, v);
              }
            else if (lua_isboolean(L, index))
              {
                bool v = lua_toboolean(L, index);
                printf("%s => %d\n", key, v);
              }
            else if (lua_isuserdata(L, index))
              {
                //                    swig_lua_userdata *usr;
                //                    swig_type_info *type;
                //                    assert(lua_isuserdata(L,index));
                //                    usr=(swig_lua_userdata*)lua_touserdata(L,index);
                //                    /* get data */
                //                    type = usr->type;
                //                    njli::AbstractFactoryObject *object =
                //                    static_cast<njli::AbstractFactoryObject*>(usr->ptr);
                //                    printf("%s => %d:%s\n", key,
                //                    object->getType(),
                //                    object->getClassName());
              }
          }
        // pop value + copy of key, leaving original key
        lua_pop(L, 2);
        // stack now contains: -1 => key; -2 => table
      }
    // stack now contains: -1 => table (when lua_next returns 0 it pops the key
    // but does not push anything.)
    // Pop table
    lua_pop(L, 1);
    // Stack is now the same as it was on entry to this function
  }

  u32 PhysicsConstraintConeTwist::type()
  {
    return JLI_OBJECT_TYPE_PhysicsConstraintConeTwist;
  }

  void PhysicsConstraintConeTwist::setNodes(Node *nodeA,
                                            const btTransform &rbAFrame)
  {
    setNodes(nodeA, NULL, rbAFrame, rbAFrame);
  }

  void PhysicsConstraintConeTwist::setNodes(Node *nodeA, Node *nodeB,
                                            const btTransform &rbAFrame,
                                            const btTransform &rbBFrame)
  {
    SDL_assert(nodeA || nodeB);

    if (nodeA && !nodeA->getPhysicsBody()->isDynamicPhysics())
      SDL_LogWarn(SDL_LOG_CATEGORY_TEST, "%s", nodeA->getName());

    if (nodeB && !nodeB->getPhysicsBody()->isDynamicPhysics())
      SDL_LogWarn(SDL_LOG_CATEGORY_TEST, "%s", nodeB->getName());

    if (nodeA)
      SDL_assert(nodeA->getPhysicsBody());

    if (nodeB)
      SDL_assert(nodeB->getPhysicsBody());

    btRigidBody *rigidBodyA = NULL;
    btRigidBody *rigidBodyB = NULL;

    if (nodeA)
      rigidBodyA = dynamic_cast<btRigidBody *>(
          nodeA->getPhysicsBody()->getCollisionObject());

    if (nodeB)
      rigidBodyB = dynamic_cast<btRigidBody *>(
          nodeB->getPhysicsBody()->getCollisionObject());

    removeConstraint();

    if (rigidBodyA && !rigidBodyB)
      {
        m_btConeTwistConstraint =
            new btConeTwistConstraint(*rigidBodyA, rbAFrame);
      }
    else if (rigidBodyA && rigidBodyB)
      {
        m_btConeTwistConstraint = new btConeTwistConstraint(
            *rigidBodyA, *rigidBodyB, rbAFrame, rbBFrame);
      }

    if (rigidBodyA)
      getNodeA()->getPhysicsBody()->setPhysicsConstraint(this);
    if (rigidBodyB)
      getNodeB()->getPhysicsBody()->setPhysicsConstraint(this);

    //        Scene *scene = njli::World::getInstance()->getScene();
    Scene *scene = getParent()->getParent()->getCurrentScene();
    scene->getPhysicsWorld()->addConstraint(this);
  }

  void PhysicsConstraintConeTwist::setAngularOnly(bool angularOnly)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setAngularOnly(angularOnly);
  }

  void PhysicsConstraintConeTwist::setLimit(int limitIndex, f32 limitValue)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setLimit(limitIndex, limitValue);
  }

  void PhysicsConstraintConeTwist::setLimit(f32 _swingSpan1, f32 _swingSpan2,
                                            f32 _twistSpan, f32 _softness,
                                            f32 _biasFactor,
                                            f32 _relaxationFactor)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setLimit(_swingSpan1, _swingSpan2, _twistSpan, _softness,
                         _biasFactor, _relaxationFactor);
  }

  const btTransform &PhysicsConstraintConeTwist::getAFrame()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getAFrame();
  }

  const btTransform &PhysicsConstraintConeTwist::getBFrame()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getBFrame();
  }

  int PhysicsConstraintConeTwist::getSolveTwistLimit()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getSolveTwistLimit();
  }

  int PhysicsConstraintConeTwist::getSolveSwingLimit()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getSolveSwingLimit();
  }

  f32 PhysicsConstraintConeTwist::getTwistLimitSign()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getTwistLimitSign();
  }

  f32 PhysicsConstraintConeTwist::getSwingSpan1()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getSwingSpan2();
  }

  f32 PhysicsConstraintConeTwist::getSwingSpan2()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getSwingSpan2();
  }

  f32 PhysicsConstraintConeTwist::getTwistSpan()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getTwistSpan();
  }

  f32 PhysicsConstraintConeTwist::getTwistAngle()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getTwistAngle();
  }

  bool PhysicsConstraintConeTwist::isPastSwingLimit()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->isPastSwingLimit();
  }

  void PhysicsConstraintConeTwist::setDamping(f32 damping)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setDamping(damping);
  }

  void PhysicsConstraintConeTwist::enableMotor(bool b)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->enableMotor(b);
  }

  void PhysicsConstraintConeTwist::setMaxMotorImpulse(f32 maxMotorImpulse)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setMaxMotorImpulse(maxMotorImpulse);
  }

  void
  PhysicsConstraintConeTwist::setMaxMotorImpulseNormalized(f32 maxMotorImpulse)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setMaxMotorImpulseNormalized(maxMotorImpulse);
  }

  f32 PhysicsConstraintConeTwist::getFixThresh()
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    return constraint->getFixThresh();
  }

  void PhysicsConstraintConeTwist::setFixThresh(f32 fixThresh)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setFixThresh(fixThresh);
  }

  void PhysicsConstraintConeTwist::setMotorTarget(const btQuaternion &q)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setMotorTarget(q);
  }

  void PhysicsConstraintConeTwist::setMotorTargetInConstraintSpace(
      const btQuaternion &q)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setMotorTargetInConstraintSpace(q);
  }

  btVector3 PhysicsConstraintConeTwist::getPointForAngle(f32 fAngleInRadians,
                                                         f32 fLength) const
  {
    const btConeTwistConstraint *constraint =
        dynamic_cast<const btConeTwistConstraint *>(getConstraint());
    return constraint->GetPointForAngle(fAngleInRadians, fLength);
  }

  void PhysicsConstraintConeTwist::setFrames(const btTransform &frameA,
                                             const btTransform &frameB)
  {
    btConeTwistConstraint *constraint =
        dynamic_cast<btConeTwistConstraint *>(getConstraint());
    constraint->setFrames(frameA, frameB);
  }

  const btTransform &PhysicsConstraintConeTwist::getFrameOffsetA() const
  {
    const btConeTwistConstraint *constraint =
        dynamic_cast<const btConeTwistConstraint *>(getConstraint());
    return constraint->getFrameOffsetA();
  }

  const btTransform &PhysicsConstraintConeTwist::getFrameOffsetB() const
  {
    const btConeTwistConstraint *constraint =
        dynamic_cast<const btConeTwistConstraint *>(getConstraint());
    return constraint->getFrameOffsetB();
  }

  btTypedConstraint *PhysicsConstraintConeTwist::getConstraint()
  {
    return m_btConeTwistConstraint;
  }

  const btTypedConstraint *PhysicsConstraintConeTwist::getConstraint() const
  {
    return m_btConeTwistConstraint;
  }

  void PhysicsConstraintConeTwist::removeConstraint()
  {
    PhysicsConstraint::removeConstraint();

    if (m_btConeTwistConstraint)
      delete m_btConeTwistConstraint;
    m_btConeTwistConstraint = NULL;
  }
}
